test/src/test/java/jenkins/model/Jenkins56575Test.java

```diff
@@ -0,0 +1,227 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import com.gargoylesoftware.htmlunit.HttpMethod;
+import com.gargoylesoftware.htmlunit.Page;
+import com.gargoylesoftware.htmlunit.WebRequest;
+import com.gargoylesoftware.htmlunit.WebResponse;
+import hudson.Launcher;
+import hudson.Main;
+import hudson.lifecycle.Lifecycle;
+import hudson.util.StreamTaskListener;
+import org.apache.commons.io.FileUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.JenkinsRule.WebClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertEquals;
+
+public class Jenkins56575Test {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    private File jar;
+
+    private AtomicReference<Boolean> restartCalled = new AtomicReference<>(false);
+    private AtomicReference<Boolean> verifyRestartableCalled = new AtomicReference<>(false);
+    private Semaphore restartSemaphore = new Semaphore(0);
+
+    private Lifecycle previousLifecycle;
+    private Field instanceField;
+
+    @Before
+    public void setupMockLifecycle() throws Exception {
+        Main.isUnitTest = false;
+        j.jenkins.setCrumbIssuer(null);
+
+        previousLifecycle = Lifecycle.get();
+
+        instanceField = Lifecycle.class.getDeclaredField("INSTANCE");
+        instanceField.setAccessible(true);
+        instanceField.set(null, new Lifecycle() {
+            @Override
+            public void verifyRestartable() {
+                verifyRestartableCalled.set(true);
+            }
+
+            @Override
+            public void restart() {
+                restartCalled.set(true);
+                restartSemaphore.release();
+            }
+        });
+    }
+
+    @Before
+    public void grabCliJar() throws IOException {
+        jar = tmp.newFile("jenkins-cli.jar");
+        FileUtils.copyURLToFile(j.jenkins.getJnlpJars("jenkins-cli.jar").getURL(), jar);
```

Comment:
I almost wonder if it's worth using a constant for these two lines. Not because the compiler won't refactor, but because it's easier for the reader to see "it's the same thing".

Reply:
I tried to turn your proposal in all shape I can imagine, I cannot find the thing you are describing :) Could you propose the "with constant" version?

Reply:
I can't figure out how to get github to make this a proper suggestion :-(

```suggestion
       String jenkinsCLI = "jenkins-cli.jar";
       jar = tmp.newFile(jenkinsCLI);
       FileUtils.copyURLToFile(j.jenkins.getJnlpJars(jenkinsCLI).getURL(), jar);
```

Reply:
Haaa ok I see your point! 
The suggestion feature is not meant for complex stuff :D